paging/guava/src/main/java/androidx/paging/ListenableFuturePagingData.kt

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2020 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.paging
+
+import androidx.annotation.CheckResult
+import androidx.concurrent.futures.await
+import com.google.common.util.concurrent.ListenableFuture
+
+/**
+ * Returns a [PagingData] containing the result of applying the given [transform] to each
+ * element, as it is loaded.
+ */
+@CheckResult
+fun <T : Any, R : Any> PagingData<T>.mapFuture(
+    transform: (T) -> ListenableFuture<R>
+): PagingData<R> = map { transform(it).await() }
+
+/**
+ * Returns a [PagingData] of all elements returned from applying the given [transform] to each
+ * element, as it is loaded.
+ */
+@CheckResult
+fun <T : Any, R : Any> PagingData<T>.flatMapFuture(
+    transform: (T) -> ListenableFuture<Iterable<R>>
+): PagingData<R> = flatMap { transform(it).await() }
+
+/**
+ * Returns a [PagingData] containing only elements matching the given [predicate].
+ */
+@CheckResult
+fun <T : Any> PagingData<T>.filterFuture(
+    predicate: (T) -> ListenableFuture<Boolean>
+): PagingData<T> = filter { predicate(it).await() }
+
+/**
+ * Returns a [PagingData] containing each original element, with an optional separator generated
+ * by [generator], given the elements before and after (or null, in boundary conditions).
+ *
+ * Note that this transform is applied asynchronously, as pages are loaded. Potential separators
+ * between pages are only computed once both pages are loaded.
+ *
+ * @sample androidx.paging.samples.insertSeparatorsSample
+ * @sample androidx.paging.samples.insertSeparatorsUiModelSample
+ */
+@CheckResult
+fun <T : R, R : Any> PagingData<T>.insertSeparatorsFuture(
```

Comment:
Aren't the params different type though? I think It should be able to disambiguate?

My reasoning is a minor improvement in discoverability, but also not a big deal if this doesn't work.

Reply:
I guess for `insertSeparators` we could use `kotlin.Pair`, the usage in Java is not as nice as in Kotlin but I think it's acceptable.

AFAIK `AsyncFunction` is currently not part of `concurrent-futures-ktx`, so using it would require a bigger change. What are your thoughts about that?

Reply:
So we talked about this internally, and the consensus was that it seems reasonable to just making `paging-guava` depend on full `guava`. `concurrent-futures` was intended to allow core modules to access LF without pulling in all of Guava, but all consumers of `paging-guava` will be using full guava anyway, and if it comes down to it we can add a callback API.

In terms of Pair, let's use a separate class instead so we can label them before / after explicitly instead of first / second. The allocation kinda sucks but we're already allocating Future anyway so :shrug: